### DISCUSSION TITLE
Howto avoid duplicates (same message from different feeds)

### DISCUSSION THREAD

**Discussion by btreut** [spacecowboy/Feeder#129] [Q&A]
If a message is found in different feeds, it is displayed multiple times. Is there any way to avoid this? I did not find an option for that.

Any hint is appreciated.

  Bernhard

**Comment by tiziodcaio** [comment]
We might think about implementing it... it's an interesting idea!

**Reply by btreut** [reply]
Thanks in advance

**Reply by spacecowboy** [reply]
who the heck are "we"?

**Reply by spacecowboy** [reply]
Please don't insinuate that you are a developer of this project @tiziodcaio

**Comment by spacecowboy** [comment]
> If a message is found in different feeds, it is displayed multiple times. Is there any way to avoid this? I did not find an option for that.

You can't avoid it. And it is a hard problem to solve.

**Comment by btreut** [comment]
> You can't avoid it. And it is a hard problem to solve.

Sorry to read this. My previous reader (Flym) did not display duplicates with (almost) the same rss feeds, but its project has stopped development more than two years ago due to some Google policy, which I don't understand, (see https://github.com/FredJul/Flym and https://web.archive.org/web/20201108022203/https://support.google.com/googleplay/android-developer/answer/10177647 ).

**Reply by spacecowboy** [reply]
Feel free to find out how they do it and describe it to me and I can take a look. But I don't have time to do code detective work for people on the Internet

**Reply by btreut** [reply]
I'll try my best. 
A first analysis: "avoid duplicates" is an option/property setting in Flym. The explanation text of the option says that it compares the title of the feed before downloading the feed and skips that feed if there exists a feed with the same title.

**Comment by btreut** [comment]
The relevant places, which compares and optionally ignores the entries with the same titles in the Flym sources are:

- Lines 427-431 in [FetcherService.kt](https://github.com/FredJul/Flym/blob/master/app/src/main/java/net/frju/flym/service/FetcherService.kt#L427)
- Lines 122-123 in [EntryDao.kt](https://github.com/FredJul/Flym/blob/master/app/src/main/java/net/frju/flym/data/dao/EntryDao.kt#L122)

but it may be more difficult with Feeder since Flym is using a SQLite DB.

**Reply by spacecowboy** [reply]
First, Feeder is also using SQLite. It's what all apps do.

Second, let's be clear about terminology. When you say feed, you mean article I believe :P

So all it's doing is comparing titles? That's pretty basic. But if that's all, then that's easy to add.

**Comment by btreut** [comment]
I used feed as a substitute for the source of an article. And yes Flym does apparently only compare the title. And yes I mixed feed up with article in my post 5 days ago, sorry for the created confusion.

E.g. in both
http://www.spiegel.de/gesundheit/index.rss
and
http://www.spiegel.de/index.rss
there is an article with title "Hebamme über Gewalt im Kreißsaal: Egal, ob die Frauen flehten, damit aufzuhören"

**Reply by spacecowboy** [reply]
So best way to implement this is to 

* Check if title is already present in the database
* If so, add a flag `hidden` to the article
* In all database queries for the UI, exclude items with `hidden` flag.

Preventing duplicates from ever saving to the database can raise issues. Consider for example 2 feeds, Feed 1 and Feed 2, which have a common article. Assume Feed 1's article is the one that the user sees.

Feed 1 might update much more often than Feed 2. So after a while, Feed 1 no longer contains the duplicate article, and Feeder deletes it from local storage as well.

Now, Feed 2's article is still in the feed file, and so on next sync it is no longer a duplicate and will suddenly appear in the UI, and the user thinks "Why is this old duplicate article suddenly here?"

Actually storing it in the DB in the first place with a `hidden` flag means Feeder won't forget that it ever WAS a duplicate and should keep hiding it.

**Comment by btreut** [comment]
Sounds good.

**Comment by btreut** [comment]
> That's pretty basic. But if that's all, then that's easy to add.

and

> So best way to implement this is to ..

do I understand right: You are plannung to add such an option?

**Reply by spacecowboy** [reply]
I'd say there's a 50% chance

**Reply by btreut** [reply]
I would really appreciate it. Thanks in advance if you can find the time to implement it.

**Comment by nvmkpk** [comment]
May be you can use additional information like the article web link. Ars Technica does cross post articles into multiple of their feeds.
